src/Common/src/TypeSystem/RuntimeDetermined/CanonUtilites.cs

```diff
@@ -0,0 +1,169 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Debug = System.Diagnostics.Debug;
+
+namespace Internal.TypeSystem
```

Comment:
Made it an algorithm but I'm pretty disgusted with the result. We're swapping perf and elegance in favor of testability of a few dozens of lines of code.